src/client/array/dc_array.c

```diff
@@ -1589,13 +1590,13 @@ dc_array_io(daos_handle_t array_oh, daos_handle_t th,
 			daos_off_t	old_array_idx;
 			daos_recx_t	*new_recxs;
 
-			iod->iod_nr++;
-
 			/** add another element to recxs */
-			D_REALLOC_ARRAY(new_recxs, iod->iod_recxs, iod->iod_nr);
+			D_REALLOC_ARRAY(new_recxs, iod->iod_recxs,
+					iod->iod_nr, iod->iod_nr + 1);
 			if (new_recxs == NULL) {
 				D_GOTO(err_stask, rc = -DER_NOMEM);
 			}
+			iod->iod_nr++;
```

Comment:
1.2.1 bug.  If realloc fails, we have erroneously incremented the value.  Not sure if it's used or not after this but it would be a problem.